File: leetcode/easy/ValidParentheses20.java
package easy;

import java.util.Stack;

// https://leetcode.com/problems/valid-parentheses/
// 20. Valid Parentheses

public class ValidParentheses20 {
    public static void main(String[] args) {
        String s = "()[]{}}";
        System.out.println(isValid(s));
    }

    public static boolean isValid(String s) {
        Stack<Character> st = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (!st.empty() && isMatching(st.peek(), c)) {
                st.pop();
            } else {
                st.push(c);
            }
        }

        return st.empty();
    }

    public static boolean isMatching(char a, char b) {
        if ((a == '(' && b == ')') || (a == '[' && b == ']') || (a == '{' && b == '}')) {
            return true;
        }
        return false;
    }

}

File: leetcode/easy/ValidParentheses20.py
# https://leetcode.com/problems/valid-parentheses/
# 20. Valid Parentheses
# '(', ')', '{', '}', '[' and ']'

class Solution:
    def isValid(self, s: str) -> bool:

        # neetcode solution

        stack = []
        close_to_open = {
            ')': '(',
            ']': '[',
            '}': '{'
        }
    
        for char in s:
            if char in close_to_open:
                if stack and stack[-1] == close_to_open[char]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(char)
            
        return not stack

        # stack = []
        # for char in s:
        #     if len(stack) == 0:
        #         stack.append(char)
        #     else:
        #         peek = stack[-1]
        #         if peek == "{" and char == "}":
        #             stack.pop()
        #         elif peek == "[" and char == "]":
        #             stack.pop()
        #         elif peek == "(" and char == ")":
        #             stack.pop()
        #         else:
        #             stack.append(char)

        # return len(stack) == 0